sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ReturnEmptyCollectionInsteadOfNull.cs

```diff
@@ -0,0 +1,121 @@
+            KnownType.System_Collections_IEnumerable,
+            KnownType.System_Array);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(ReportIfReturnsNull,
+                SyntaxKind.MethodDeclaration, SyntaxKind.PropertyDeclaration);
+        }
+
+        private static void ReportIfReturnsNull(SyntaxNodeAnalysisContext context)
+        {
+            if (!IsReturningCollection(context))
+            {
+                return;
+            }
+
+            var propertyDeclaration = context.Node as PropertyDeclarationSyntax;
+            var methodDeclaration = context.Node as MethodDeclarationSyntax;
+
+            var expressionBody = propertyDeclaration != null
+                ? propertyDeclaration.ExpressionBody
+                : methodDeclaration.ExpressionBody;
+            if (expressionBody != null)
+            {
+                var arrowedNullLiteral = GetNullLiteralOrDefault(expressionBody);
+                if (arrowedNullLiteral != null)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(rule, arrowedNullLiteral.GetLocation()));
+                }
+
+                return;
+            }
+
+            var methodBody = propertyDeclaration?.AccessorList.Accessors
+                .FirstOrDefault(accessor => accessor.IsKind(SyntaxKind.GetAccessorDeclaration))?.Body
+                ?? methodDeclaration?.Body;
+            if (methodBody == null)
+            {
+                return;
+            }
+
+            var returnNullStatements = GetReturnNullStatements(methodBody)
+                .Select(returnStatement => returnStatement.GetLocation())
+                .ToList();
+            if (returnNullStatements.Count > 0)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(rule, returnNullStatements[0],
+                    additionalLocations: returnNullStatements.Skip(1)));
+            }
+        }
+
+        private static bool IsReturningCollection(SyntaxNodeAnalysisContext context)
+        {
+            var symbol = context.SemanticModel.GetDeclaredSymbol(context.Node);
+
+            var methodSymbol = (symbol as IPropertySymbol)?.GetMethod ?? symbol as IMethodSymbol;
+
+            return methodSymbol != null &&
+                !methodSymbol.ReturnType.Is(KnownType.System_String) &&
+                methodSymbol.ReturnType.DerivesOrImplementsAny(CollectionTypes);
+        }
+
+        private static LiteralExpressionSyntax GetNullLiteralOrDefault(ArrowExpressionClauseSyntax expressionBody)
+        {
+            return expressionBody.Expression.IsKind(SyntaxKind.NullLiteralExpression)
+                ? (LiteralExpressionSyntax)expressionBody.Expression
+                : null;
+        }
+
+        private static IEnumerable<ReturnStatementSyntax> GetReturnNullStatements(BlockSyntax methodBlock)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Consider using more general type 'Microsoft.CodeAnalysis.SyntaxNode' instead of 'Microsoft.CodeAnalysis.CSharp.Syntax.BlockSyntax'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3242)